app/controllers/concerns/rendering.rb

```diff
@@ -11,10 +11,15 @@ def render_json(model, **args)
       render({ json: serializer.new(model, opts) }.merge(args))
     end
 
-    def render_error(models, **args)
+    def render_error(messages_or_models, status: :not_acceptable, **opts)
+      messages = [messages_or_models].flatten
+      messages = messages.flat_map do |msg_or_model|
+        msg_or_model.respond_to?(:errors) ? model_errors(msg_or_model) : msg_or_model
+      end
```

Comment:
I'm not a huge fan of type checking like this. We could avoid it by simply using keyword arguments:

```suggestion
def render_error(messages: [], models: [], status: :not_acceptable, **args)
  messages ||= model_errors(models)
  ...
end
```

Also, `model_errors` is only used here, I think. So if you decide not to use it for some reason, we should remove it.

Reply:
I wanted to avoid changing the signature, as it is already being used for model errors. This is a duck-typing (checking an "interface").

Reply:
Duck typing is allowing different types to respond to the same message/method without explicit type checks. Like if you had a Table class and a Cat class and both respond to `number_of_legs`. When you write `respond_to?` that's an explicit type check which defeats the purpose of duck typing (the purpose being to hide complexity and share code).

This method is only used in the profiles controller, so it would be trivial to change the signature.

Reply:
I've reworked it by splitting it to a different method. Check out first commit (c192fef).

`responds_to?` is a feature/interface check (like the `number_of_legs` from your example). I don't want to dispute your interpretation of duck-typing. Similar principle is for example used in Rails. See [`acts_like?` implementation](https://github.com/rails/rails/blob/b2eb1d1c55a59fee1e6c4cba7030d8ceb524267c/activesupport/lib/active_support/core_ext/object/acts_like.rb) for inspiration.

I guess your concern was more to the fact that there were possible features checked. One that if it had `errors` and second if it hadn't.

(Technically the check was incomplete, as the subsequent method read `full_messages` out of `errors`.)